hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AMetadataPersistenceException.java

```diff
@@ -0,0 +1,115 @@
+ */
+
+package org.apache.hadoop.fs.s3a;
+
+import static org.apache.hadoop.test.LambdaTestUtils.intercept;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doThrow;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.s3a.s3guard.MetadataStore;
+import org.apache.hadoop.fs.s3a.s3guard.PathMetadata;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.Mockito;
+
+/**
+ * Tests failed writes to metadata store generate the expected
+ * MetadataPersistenceException.
+ */
+@RunWith(Parameterized.class)
+public class ITestS3AMetadataPersistenceException extends AbstractS3ATestBase {
+  private S3AFileSystem fs;
+  private MetadataStore originalMetadataStore;
+  private IOException ioException;
+  private final boolean failOnError;
+
+  public ITestS3AMetadataPersistenceException(boolean failOnError) {
+    this.failOnError = failOnError;
+  }
+
+  @Parameterized.Parameters
+  public static Collection<Object[]> params() {
+    return Arrays.asList(new Object[][]{
+        {true},
+        {false}
+    });
+  }
+
+  @Override
+  protected Configuration createConfiguration() {
+    Configuration conf = super.createConfiguration();
+    conf.set(Constants.FAIL_ON_METADATA_WRITE_ERROR,
+        Boolean.toString(failOnError));
+    return conf;
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    fs = getFileSystem();
+    originalMetadataStore = fs.getMetadataStore();
+    ioException = new IOException();
+  }
+
+  @After
+  public void tearDown() {
+    fs.setMetadataStore(originalMetadataStore);
+  }
+
+  private void useIOExceptionMetadataStore() throws IOException {
+    MetadataStore metadataStoreSpy = Mockito.spy(originalMetadataStore);
+    doThrow(ioException).when(metadataStoreSpy)
+        .put(any(PathMetadata.class));
+    fs.setMetadataStore(metadataStoreSpy);
+  }
+
+  @Test
+  public void testFailedMetadataUpdate() throws Throwable {
+    // write a trivial file
+    Path testFile = path("testFile");
+    FSDataOutputStream outputStream = fs.create(testFile);
+    outputStream.write(1);
+
+    useIOExceptionMetadataStore();
+
+    if (failOnError) {
+      // close should throw the expected exception
+      Exception thrown = intercept(MetadataPersistenceException.class, () -> {
+        outputStream.close();
+      });
+      assertTrue("expected MetadataPersistenceException wasn't thrown",
+          thrown instanceof MetadataPersistenceException);
+      assertEquals("cause didn't match original exception",
+          ioException, thrown.getCause());
+    } else {
+      // close should merely log and increment the statistic
+      outputStream.close();
+      assertEquals(1,
```

Comment:
this would be brittle to any existing failures in the test run. It will probably be moot once you switch to a new instance, but you shoud use `S3ATestUtils.MetricDiff` for your asserts

Reply:
Thanks for the tip.  Updated in https://github.com/apache/hadoop/pull/666/commits/8ff542741ab3386a9ce967fb2f36fab2265413f4